x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,506 @@
+    add_new_channels, try_add_channel, try_remove_channel, try_reset_channel_quota,
+};
+use crate::msg::{ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg};
+use crate::state::{ChannelFlow, FlowType, CHANNEL_FLOWS, GOVMODULE, IBCMODULE};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:rate-limiter";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    IBCMODULE.save(deps.storage, &msg.ibc_module)?;
+    GOVMODULE.save(deps.storage, &msg.gov_module)?;
+
+    add_new_channels(deps, msg.channels, env.block.time)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("ibc_module", msg.ibc_module.to_string())
+        .add_attribute("gov_module", msg.gov_module.to_string()))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::SendPacket {
+            channel_id,
+            channel_value,
+            funds,
+        } => try_transfer(
+            deps,
+            info.sender,
+            channel_id,
+            channel_value,
+            funds,
+            FlowType::Out,
+            env.block.time,
+        ),
+        ExecuteMsg::RecvPacket {
+            channel_id,
+            channel_value,
+            funds,
+        } => try_transfer(
+            deps,
+            info.sender,
+            channel_id,
+            channel_value,
+            funds,
+            FlowType::In,
+            env.block.time,
+        ),
+        ExecuteMsg::AddChannel { channel_id, quotas } => {
+            try_add_channel(deps, info.sender, channel_id, quotas, env.block.time)
+        }
+        ExecuteMsg::RemoveChannel { channel_id } => {
+            try_remove_channel(deps, info.sender, channel_id)
+        }
+        ExecuteMsg::ResetChannelQuota {
+            channel_id,
+            quota_id,
+        } => try_reset_channel_quota(deps, info.sender, channel_id, quota_id, env.block.time),
+    }
+}
+
+pub struct ChannelFlowResponse {
+    pub channel_flow: ChannelFlow,
+    pub used: u128,
+    pub max: u128,
+}
+
+pub fn try_transfer(
+    deps: DepsMut,
+    sender: Addr,
+    channel_id: String,
+    channel_value: u128,
+    funds: u128,
+    direction: FlowType,
+    now: Timestamp,
+) -> Result<Response, ContractError> {
+    // Only the IBCMODULE can execute transfers
+    let ibc_module = IBCMODULE.load(deps.storage)?;
+    if sender != ibc_module {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let channels = CHANNEL_FLOWS.may_load(deps.storage, &channel_id)?;
```

Comment:
```suggestion
    let channel_flow_trackers = CHANNEL_FLOWS.may_load(deps.storage, &channel_id)?;
```

I was a bit confused about that this was iterating over IBC channels on first pass

Reply:
this makes sense. Had a hard time naming that one, but I like "trackers". I'll update that function

Reply:
Changed a few of the names. Marking this as resolved for now